llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp

```diff
@@ -340,8 +340,13 @@ Value *VPInstruction::generateInstruction(VPTransformState &State,
       auto *Phi = State.get(getOperand(0), 0);
       // The loop step is equal to the vectorization factor (num of SIMD
       // elements) times the unroll factor (num of SIMD instructions).
-      Value *Step =
-          createStepForVF(Builder, Phi->getType(), State.VF, State.UF);
+      Value *Step;
+      {
+        BasicBlock *VectorPH = State.CFG.getPreheaderBBFor(this);
+        IRBuilder<>::InsertPointGuard Guard(Builder);
+        Builder.SetInsertPoint(VectorPH->getTerminator());
```

Comment:
This is fine. Can alternatively define a separate builder set to the vector preheader to serve createStepForVF, independent of `Builder` which serves CreateAdd.

Perhaps worth a comment that step for VF must be loop invariant, and for scalable VF's generates a call to vscale which should be placed in the preheader.

Reply:
Adjusted in the latest version, thanks!